internal/eventlog/cursor/cursor.go

```diff
@@ -0,0 +1,100 @@
+// Package cursor implements time-ordered item cursors for an event log.
+package cursor
+
+import (
+	"errors"
+	"fmt"
+	"strconv"
+	"strings"
+	"time"
+)
+
+// A Source produces cursors based on a time index generator and a sequence
+// counter. A zero-valued Source is ready for use with defaults as described.
+type Source struct {
+	// This function is called to produce the current time index.
+	// If nil, it defaults to time.Now().UnixNano().
+	TimeIndex func() int64
+
+	// The current counter value used for sequence number generation.  It is
+	// incremented in-place each time a cursor is generated.
+	Counter int64
+}
+
+func (s *Source) timeIndex() int64 {
+	if s.TimeIndex == nil {
+		return time.Now().UnixNano()
+	}
+	return s.TimeIndex()
+}
+
+func (s *Source) nextCounter() int64 {
+	s.Counter++
+	return s.Counter
+}
+
+// Cursor produces a fresh cursor from s at the current time index and counter.
+func (s *Source) Cursor() Cursor {
+	return Cursor{
+		timestamp: uint64(s.timeIndex()),
+		sequence:  uint16(s.nextCounter() & 0xffff),
```

Comment:
Yes—since the sequence number is only 16 bits, this just wraps the counter to that size. We could just make both 64 bits wide—but we don't really need that many bits. This is just a safety condition in case (say) the system clock only resolves milliseconds and we got a few events delivered during the same ms period. (The units are nanoseconds, but the clock may not have that much resolution)

Reply:
Ah neat. Do all events emitted when a block is committed have the same timestamp?

Reply:
Ideally no—the timestamp is assigned by the `cursor.Source`. By default that calls `time.Now().UnixNano()`, but it is possible that two calls may report the same value.

It's very unlikely that two events occur at the same nanosecond, but if the system clock only reports milli- or micro-seconds, it could happen. So the sequence number basically includes a value that is _not_ time dependent. 

We need some measure of time to deal with the time window, so we can't just make it a pure sequence number. But we also want to avoid having collisions on time alone, even though those should be unlikely.